### PR TITLE
Provide a module type signature and a connect_device to be able to create a Mirage device

### PR DESCRIPTION
**Pull request by dinosaure**
In the same view than mirage/ocaml-dns#297 and in the goal of mirage/mirage#1291, this PR wants to provide:
- a simple interface which can be use by the `mirage`/`functoria` tool
- a composition between `Dns_client_mirage.S`/`Dns_client_mirage.t` and this module
- the `connect_device` function which will be used by the `mirage` tool to create an `Happy_eyeballs_mirage.t`

You probably are a bit afraid by a constraint: `Dns_client_mirage.S with type Transport.stack = S.t` - for my point of view, it's completely fine. The `functoria` DSL does not have the ability to unify types/devices and if the user uses an other implementation of `Tcpip.Stack..V4V6` than the one used to make the `Dns_client_mirage.S` module, OCaml will complain about incompatible type. As far as I can say, we are not able to constraint that _via_ the `functoria` DSL.

However, I really think that it should help us to define our final `Mirage.happy_eyeballs` (& our `Mirage.dns_client`). The last goal is to use them to create an `mimic_happy_eyeballs` which can be used then by Git devices.

### CODE REVIEW

**Comment by dinosaure** [comment]
I took the opportunity to open `Msg of string` errors to be more composable with something else.

**Comment by dinosaure** [comment]
Finally, I fixed constraints to be able to build a `mimic` device with a possible `dns` and a `happy-eyeballs` devices. However, I would like to say that it was very difficult to figure out how to fix these constraints when a `module type of Dns_client.Make(Transport)` into signatures generates new types without constraints along projects.

It's a more systemic problem but I really think that in the purpose of `mirage`/`functoria`, we should not generate a signature from a _functor_.

**Comment by dinosaure** [comment]
**NOTE**: by design, `connect_device` should **not** appear into the signature but only into the _functor_. I put it but if you are more confident to put it outside (as any devices in `mirage`), I can do it.

**Comment by dinosaure** [comment]
As far as I can say, I just moved `connect_device` outside the signature `S` (as we usually do for MirageOS _device_) but nothing should be upgraded then from `dns.6.2.0`. So, you are able to review this PR :+1:. Then, we will able to propose a device into `mirage/mirage`.

**Comment by dinosaure** [comment]
I finally did the `mimic` layer (expected by `git` and `paf`) here: https://github.com/dinosaure/mimic/pull/11 which will be compatible with MirageOS.

File: mirage/happy_eyeballs_mirage.mli [hunk 1]
```diff
@@ -1,17 +1,40 @@
-module Make (R : Mirage_random.S) (T : Mirage_time.S) (C : Mirage_clock.MCLOCK) (P : Mirage_clock.PCLOCK) (S : Tcpip.Stack.V4V6) : sig
-  module DNS : module type of Dns_client_mirage.Make(R)(T)(C)(P)(S)
```
Comment (hannesm) [inline mirage/happy_eyeballs_mirage.mli:2]: Somehow removing this functor application makes everything so complex -- now there's a "type dns" and "type flow" in the interface, and some strange "module Transport... with type io_addr = ... and type +'a io ...". Isn't there a way around it? Is the issue that OCaml's functors are generative (which can be avoided by https://discuss.ocaml.org/t/generative-functors/2945)?

**Comment by hannesm** [comment]
I did not look deep into the implementation, but I don't think adding abstract types to the interface is the way to go. It'd be nice if we could find a path without that noise.

File: mirage/happy_eyeballs_mirage.mli (hunk 1 above)
Comment (dinosaure) [reply mirage/happy_eyeballs_mirage.mli:2]: The main issue is to pass the `DNS.t` to the `connect` function. An other solution should be to create inside the `connect`  the expected `DNS.t` with a given `nameserver` argument (as `Dns_client_mirage`) but that mostly mean that: we don't create the `happy_eyeballs` device from a DNS device but from nothing. Are you ok with that?

Comment (hannesm) [reply mirage/happy_eyeballs_mirage.mli:2]: I see what you mean and need to think about it. At the moment it feels like the approach in this PR is fine, and apart from the type aliases (and implied complexity) provides us with a nice set of tradeoffs.

**Comment by hannesm** [comment]
Thanks. I merged this PR manually after minor code style changes.

**Comment by dinosaure** [comment]
Thanks, it unlock me several projects then, I will continue the stuff on the MirageOS side :+1:.

**Pull request closed without merging**